File: lang/kotlin/cxlib/src/main/kotlin/cx/CXPath.kt
package cx

// ── Predicates ────────────────────────────────────────────────────────────────

sealed class CXPred
data class CXPredAttrExists(val attr: String) : CXPred()
data class CXPredAttrCmp(val attr: String, val op: String, val `val`: Any?) : CXPred()
data class CXPredChildExists(val name: String) : CXPred()
data class CXPredNot(val inner: CXPred) : CXPred()
data class CXPredBoolAnd(val left: CXPred, val right: CXPred) : CXPred()
data class CXPredBoolOr(val left: CXPred, val right: CXPred) : CXPred()
data class CXPredPosition(val pos: Int = 0, val isLast: Boolean = false) : CXPred()
data class CXPredFuncContains(val attr: String, val `val`: String) : CXPred()
data class CXPredFuncStartsWith(val attr: String, val `val`: String) : CXPred()

enum class CXAxis { CHILD, DESCENDANT }

data class CXStep(val axis: CXAxis, val name: String, val preds: List<CXPred>)
data class CXPathExpr(val steps: List<CXStep>)


// ── Lexer ─────────────────────────────────────────────────────────────────────

private class Lexer(val src: String) {
    var pos: Int = 0

    fun skipWs() {
        while (pos < src.length && src[pos] == ' ') pos++
    }

    fun peekStr(s: String): Boolean = src.startsWith(s, pos)

    fun eatStr(s: String): Boolean {
        if (peekStr(s)) { pos += s.length; return true }
        return false
    }

    fun eatChar(c: Char): Boolean {
        if (pos < src.length && src[pos] == c) { pos++; return true }
        return false
    }

    fun readIdent(): String {
        val start = pos
        while (pos < src.length) {
            val c = src[pos]
            if (c.isLetterOrDigit() || c in "_-.:%") pos++ else break
        }
        return src.substring(start, pos)
    }

    fun readQuoted(): String {
        if (!eatChar('\'')) throw IllegalArgumentException("CXPath parse error: expected ' at pos $pos  expr: $src")
        val start = pos
        while (pos < src.length && src[pos] != '\'') pos++
        val s = src.substring(start, pos)
        if (!eatChar('\'')) throw IllegalArgumentException("CXPath parse error: unterminated string at pos $pos  expr: $src")
        return s
    }
}


// ── Parser ────────────────────────────────────────────────────────────────────

fun cxpathParse(expr: String): CXPathExpr {
    val l = Lexer(expr)
    val steps = parseSteps(l)
    if (l.pos != l.src.length)
        throw IllegalArgumentException("CXPath parse error: unexpected characters at pos ${l.pos}  expr: $expr")
    if (steps.isEmpty())
        throw IllegalArgumentException("CXPath parse error: empty expression  expr: $expr")
    return CXPathExpr(steps)
}

private fun parseSteps(l: Lexer): List<CXStep> {
    val steps = mutableListOf<CXStep>()
    val axis = when {
        l.peekStr("//") -> { l.pos += 2; CXAxis.DESCENDANT }
        l.peekStr("/")  -> { l.pos += 1; CXAxis.CHILD }
        else -> CXAxis.CHILD
    }
    steps.add(parseOneStep(l, axis))
    while (true) {
        l.skipWs()
        when {
            l.peekStr("//") -> { l.pos += 2; steps.add(parseOneStep(l, CXAxis.DESCENDANT)) }
            l.peekStr("/")  -> { l.pos += 1; steps.add(parseOneStep(l, CXAxis.CHILD)) }
            else -> break
        }
    }
    return steps
}

private fun parseOneStep(l: Lexer, axis: CXAxis): CXStep {
    l.skipWs()
    val name = if (l.eatChar('*')) {
        ""
    } else {
        val n = l.readIdent()
        if (n.isEmpty()) throw IllegalArgumentException("CXPath parse error: expected element name at pos ${l.pos}  expr: ${l.src}")
        n
    }
    val preds = mutableListOf<CXPred>()
    while (true) {
        l.skipWs()
        if (l.peekStr("[")) preds.add(parsePredBracket(l))
        else break
    }
    return CXStep(axis, name, preds)
}

private fun parsePredBracket(l: Lexer): CXPred {
    if (!l.eatChar('[')) throw IllegalArgumentException("CXPath parse error: expected [ at pos ${l.pos}  expr: ${l.src}")
    l.skipWs()
    val pred = parsePredExpr(l)
    l.skipWs()
    if (!l.eatChar(']')) throw IllegalArgumentException("CXPath parse error: expected ] at pos ${l.pos}  expr: ${l.src}")
    return pred
}

private fun parsePredExpr(l: Lexer): CXPred {
    val left = parsePredTerm(l)
    l.skipWs()
    val saved = l.pos
    val word = l.readIdent()
    if (word == "or") {
        l.skipWs()
        val right = parsePredTerm(l)
        return CXPredBoolOr(left, right)
    }
    l.pos = saved
    return left
}

private fun parsePredTerm(l: Lexer): CXPred {
    val left = parsePredFactor(l)
    l.skipWs()
    val saved = l.pos
    val word = l.readIdent()
    if (word == "and") {
        l.skipWs()
        val right = parsePredFactor(l)
        return CXPredBoolAnd(left, right)
    }
    l.pos = saved
    return left
}

private fun parsePredFactor(l: Lexer): CXPred {
    l.skipWs()
    // not(...)
    if (l.peekStr("not(") || l.peekStr("not (")) {
        l.readIdent()  // consume 'not'
        l.skipWs()
        if (!l.eatChar('(')) throw IllegalArgumentException("CXPath parse error: expected ( after not  expr: ${l.src}")
        l.skipWs()
        val inner = parsePredExpr(l)
        l.skipWs()
        if (!l.eatChar(')')) throw IllegalArgumentException("CXPath parse error: expected ) after not(...)  expr: ${l.src}")
        return CXPredNot(inner)
    }
    // contains(@attr, val)
    if (l.peekStr("contains(")) {
        l.readIdent()  // consume 'contains'
        l.skipWs()
        if (!l.eatChar('(')) throw IllegalArgumentException("CXPath parse error: expected ( after contains  expr: ${l.src}")
        l.skipWs()
        if (!l.eatChar('@')) throw IllegalArgumentException("CXPath parse error: expected @attr in contains()  expr: ${l.src}")
        val attr = l.readIdent()
        l.skipWs()
        if (!l.eatChar(',')) throw IllegalArgumentException("CXPath parse error: expected , in contains()  expr: ${l.src}")
        l.skipWs()
        val `val` = parseScalarStr(l)
        l.skipWs()
        if (!l.eatChar(')')) throw IllegalArgumentException("CXPath parse error: expected ) after contains(...)  expr: ${l.src}")
        return CXPredFuncContains(attr, `val`)
    }
    // starts-with(@attr, val)
    if (l.peekStr("starts-with(")) {
        while (l.pos < l.src.length && l.src[l.pos] != '(') l.pos++
        if (!l.eatChar('(')) throw IllegalArgumentException("CXPath parse error: expected ( after starts-with  expr: ${l.src}")
        l.skipWs()
        if (!l.eatChar('@')) throw IllegalArgumentException("CXPath parse error: expected @attr in starts-with()  expr: ${l.src}")
        val attr = l.readIdent()
        l.skipWs()
        if (!l.eatChar(',')) throw IllegalArgumentException("CXPath parse error: expected , in starts-with()  expr: ${l.src}")
        l.skipWs()
        val `val` = parseScalarStr(l)
        l.skipWs()
        if (!l.eatChar(')')) throw IllegalArgumentException("CXPath parse error: expected ) after starts-with(...)  expr: ${l.src}")
        return CXPredFuncStartsWith(attr, `val`)
    }
    // last()
    if (l.peekStr("last()")) {
        l.pos += 6
        return CXPredPosition(isLast = true)
    }
    // (grouped expr)
    if (l.peekStr("(")) {
        l.eatChar('(')
        l.skipWs()
        val inner = parsePredExpr(l)
        l.skipWs()
        if (!l.eatChar(')')) throw IllegalArgumentException("CXPath parse error: expected ) at pos ${l.pos}  expr: ${l.src}")
        return inner
    }
    // @attr comparison or existence
    if (l.pos < l.src.length && l.src[l.pos] == '@') {
        l.eatChar('@')
        val attr = l.readIdent()
        l.skipWs()
        val op = parseOp(l)
        if (op.isEmpty()) return CXPredAttrExists(attr)
        l.skipWs()
        val `val` = parseScalarVal(l)
        return CXPredAttrCmp(attr, op, `val`)
    }
    // integer position predicate
    if (l.pos < l.src.length && l.src[l.pos].isDigit()) {
        val start = l.pos
        while (l.pos < l.src.length && l.src[l.pos].isDigit()) l.pos++
        return CXPredPosition(pos = l.src.substring(start, l.pos).toInt())
    }
    // bare name → child existence
    val name = l.readIdent()
    if (name.isNotEmpty()) return CXPredChildExists(name)
    throw IllegalArgumentException("CXPath parse error: unexpected character at pos ${l.pos}  expr: ${l.src}")
}

private fun parseOp(l: Lexer): String {
    for (op in listOf("!=", ">=", "<=", "=", ">", "<")) {
        if (l.eatStr(op)) return op
    }
    return ""
}

private fun autotypeValue(s: String): Any? {
    if (s == "true")  return true
    if (s == "false") return false
    if (s == "null")  return null
    s.toLongOrNull()?.let { return it }
    s.toDoubleOrNull()?.let { return it }
    return s
}

private fun parseScalarVal(l: Lexer): Any? {
    if (l.peekStr("'")) return l.readQuoted()
    val s = l.readIdent()
    if (s.isEmpty()) throw IllegalArgumentException("CXPath parse error: expected value at pos ${l.pos}  expr: ${l.src}")
    return autotypeValue(s)
}

private fun parseScalarStr(l: Lexer): String {
    if (l.peekStr("'")) return l.readQuoted()
    return l.readIdent()
}


// ── Evaluator ─────────────────────────────────────────────────────────────────

fun collectStep(ctx: Element, expr: CXPathExpr, stepIdx: Int, result: MutableList<Element>) {
    if (stepIdx >= expr.steps.size) return
    val step = expr.steps[stepIdx]
    if (step.axis == CXAxis.CHILD) {
        val candidates = ctx.items.filterIsInstance<Element>()
            .filter { step.name.isEmpty() || it.name == step.name }
        for ((i, child) in candidates.withIndex()) {
            if (predsMatch(child, step.preds, candidates, i)) {
                if (stepIdx == expr.steps.size - 1) {
                    result.add(child)
                } else {
                    collectStep(child, expr, stepIdx + 1, result)
                }
            }
        }
    } else {
        collectDescendants(ctx, expr, stepIdx, result)
    }
}

private fun collectDescendants(ctx: Element, expr: CXPathExpr, stepIdx: Int, result: MutableList<Element>) {
    val step = expr.steps[stepIdx]
    val isLast = stepIdx == expr.steps.size - 1
    val candidates = ctx.items.filterIsInstance<Element>()
        .filter { step.name.isEmpty() || it.name == step.name }
    for ((i, child) in candidates.withIndex()) {
        if (predsMatch(child, step.preds, candidates, i)) {
            if (isLast) {
                result.add(child)
            } else {
                collectStep(child, expr, stepIdx + 1, result)
            }
        }
        // Always recurse deeper for descendant axis
        collectDescendants(child, expr, stepIdx, result)
    }
    // Also descend into non-matching children for named steps
    if (step.name.isNotEmpty()) {
        for (child in ctx.items) {
            if (child is Element && child.name != step.name) {
                collectDescendants(child, expr, stepIdx, result)
            }
        }
    }
}


// ── Predicate evaluators ──────────────────────────────────────────────────────

private fun predsMatch(el: Element, preds: List<CXPred>, siblings: List<Element>, idx: Int): Boolean =
    preds.all { predEval(el, it, siblings, idx) }

private fun predEval(el: Element, pred: CXPred, siblings: List<Element>, idx: Int): Boolean = when (pred) {
    is CXPredAttrExists -> el.attr(pred.attr) != null
    is CXPredAttrCmp -> {
        val v = el.attr(pred.attr)
        if (v == null) false else compare(v, pred.op, pred.`val`)
    }
    is CXPredChildExists -> el.get(pred.name) != null
    is CXPredNot -> !predEval(el, pred.inner, siblings, idx)
    is CXPredBoolAnd -> predEval(el, pred.left, siblings, idx) && predEval(el, pred.right, siblings, idx)
    is CXPredBoolOr  -> predEval(el, pred.left, siblings, idx) || predEval(el, pred.right, siblings, idx)
    is CXPredPosition -> if (pred.isLast) idx == siblings.size - 1 else idx == pred.pos - 1
    is CXPredFuncContains -> {
        val v = el.attr(pred.attr)
        v != null && pred.`val` in valToStr(v)
    }
    is CXPredFuncStartsWith -> {
        val v = el.attr(pred.attr)
        v != null && valToStr(v).startsWith(pred.`val`)
    }
}

private fun valToStr(v: Any?): String = when {
    v == null -> "null"
    v is Boolean -> if (v) "true" else "false"
    else -> v.toString()
}

private fun scalarEq(a: Any?, b: Any?): Boolean {
    // Guard: bool vs non-bool cross-type must never match (unlike Python where bool < int)
    val aBool = a is Boolean
    val bBool = b is Boolean
    if (aBool != bBool) return false
    // Numeric equality via Double (handles Long vs Double comparisons)
    if (a is Number && b is Number)
        return a.toDouble() == b.toDouble()
    return a == b
}

private fun compare(actual: Any?, op: String, expected: Any?): Boolean = when (op) {
    "="  -> scalarEq(actual, expected)
    "!=" -> !scalarEq(actual, expected)
    else -> {
        val a = toF64(actual)
        val b = toF64(expected)
        when (op) {
            ">"  -> a > b
            "<"  -> a < b
            ">=" -> a >= b
            "<=" -> a <= b
            else -> false
        }
    }
}

private fun toF64(v: Any?): Double {
    if (v is Boolean) throw IllegalArgumentException("CXPath: numeric comparison requires numeric value, got bool: $v")
    if (v is Number) return v.toDouble()
    throw IllegalArgumentException("CXPath: numeric comparison requires numeric attribute value, got: $v")
}


// ── cxpathElemMatches (for transformAll) ──────────────────────────────────────

fun cxpathElemMatches(el: Element, expr: CXPathExpr): Boolean {
    if (expr.steps.isEmpty()) return false
    val last = expr.steps.last()
    if (last.name.isNotEmpty() && last.name != el.name) return false
    val nonPos = last.preds.filter { it !is CXPredPosition }
    return predsMatch(el, nonPos, emptyList(), 0)
}


// ── Transform helpers ─────────────────────────────────────────────────────────

fun elemDetached(e: Element): Element =
    Element(
        name = e.name,
        anchor = e.anchor,
        merge = e.merge,
        dataType = e.dataType,
        attrs = e.attrs.map { Attr(it.name, it.value, it.dataType) }.toMutableList(),
        items = e.items.toMutableList(),
    )

fun docReplaceAt(d: CXDocument, idx: Int, el: Element): CXDocument {
    val newElements = d.elements.mapIndexed { i, n -> if (i == idx) el else n }.toMutableList()
    return CXDocument(newElements, d.prolog.toMutableList(), d.doctype)
}

fun elemReplaceItemAt(e: Element, idx: Int, child: Node): Element =
    Element(
        name = e.name,
        anchor = e.anchor,
        merge = e.merge,
        dataType = e.dataType,
        attrs = e.attrs,
        items = e.items.mapIndexed { i, n -> if (i == idx) child else n }.toMutableList(),
    )

fun pathCopyElement(e: Element, parts: List<String>, f: (Element) -> Element): Element? {
    for ((i, item) in e.items.withIndex()) {
        if (item is Element && item.name == parts[0]) {
            return if (parts.size == 1) {
                elemReplaceItemAt(e, i, f(elemDetached(item)))
            } else {
                val updated = pathCopyElement(item, parts.drop(1), f)
                if (updated != null) elemReplaceItemAt(e, i, updated) else null
            }
        }
    }
    return null
}

fun rebuildNode(node: Node, expr: CXPathExpr, f: (Element) -> Element): Node {
    if (node !is Element) return node
    val newItems = node.items.map { rebuildNode(it, expr, f) }.toMutableList()
    val newEl = Element(
        name = node.name,
        anchor = node.anchor,
        merge = node.merge,
        dataType = node.dataType,
        attrs = node.attrs,
        items = newItems,
    )
    return if (cxpathElemMatches(newEl, expr)) f(elemDetached(newEl)) else newEl
}

File: lang/python/cxlib/cxpath.py
"""CXPath parser, evaluator, and transform helpers."""
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any, Callable, Optional


# ── CXPath AST ────────────────────────────────────────────────────────────────

@dataclass
class CXPredAttrExists:
    attr: str

@dataclass
class CXPredAttrCmp:
    attr: str
    op: str
    val: Any

@dataclass
class CXPredChildExists:
    name: str

@dataclass
class CXPredNot:
    inner: Any  # CXPred

@dataclass
class CXPredBoolAnd:
    left: Any
    right: Any

@dataclass
class CXPredBoolOr:
    left: Any
    right: Any

@dataclass
class CXPredPosition:
    pos: int = 0
    is_last: bool = False

@dataclass
class CXPredFuncContains:
    attr: str
    val: str

@dataclass
class CXPredFuncStartsWith:
    attr: str
    val: str

@dataclass
class _CXStep:
    axis: str   # 'child' | 'descendant'
    name: str   # '' = wildcard (*)
    preds: list = field(default_factory=list)

@dataclass
class _CXPathExpr:
    steps: list = field(default_factory=list)


# ── Tokenizer ─────────────────────────────────────────────────────────────────

class _Lexer:
    def __init__(self, src: str):
        self.src = src
        self.pos = 0

    def skip_ws(self):
        while self.pos < len(self.src) and self.src[self.pos] == ' ':
            self.pos += 1

    def peek_str(self, s: str) -> bool:
        return self.src[self.pos:].startswith(s)

    def eat_str(self, s: str) -> bool:
        if self.peek_str(s):
            self.pos += len(s)
            return True
        return False

    def eat_char(self, c: str) -> bool:
        if self.pos < len(self.src) and self.src[self.pos] == c:
            self.pos += 1
            return True
        return False

    def read_ident(self) -> str:
        start = self.pos
        while self.pos < len(self.src):
            c = self.src[self.pos]
            if c.isalnum() or c in '_-.:%':
                self.pos += 1
            else:
                break
        return self.src[start:self.pos]

    def read_quoted(self) -> str:
        if not self.eat_char("'"):
            raise ValueError(f"CXPath parse error: expected ' at pos {self.pos}  expr: {self.src}")
        start = self.pos
        while self.pos < len(self.src) and self.src[self.pos] != "'":
            self.pos += 1
        s = self.src[start:self.pos]
        if not self.eat_char("'"):
            raise ValueError(f"CXPath parse error: unterminated string at pos {self.pos}  expr: {self.src}")
        return s


# ── Parser ────────────────────────────────────────────────────────────────────

def cxpath_parse(expr: str) -> _CXPathExpr:
    l = _Lexer(expr)
    steps = _parse_steps(l)
    if l.pos != len(l.src):
        raise ValueError(f"CXPath parse error: unexpected characters at pos {l.pos}  expr: {expr}")
    if not steps:
        raise ValueError(f"CXPath parse error: empty expression  expr: {expr}")
    return _CXPathExpr(steps=steps)


def _parse_steps(l: _Lexer) -> list:
    steps = []
    axis = 'child'
    if l.peek_str('//'):
        l.pos += 2
        axis = 'descendant'
    elif l.peek_str('/'):
        l.pos += 1
        axis = 'child'
    steps.append(_parse_one_step(l, axis))
    while True:
        l.skip_ws()
        if l.peek_str('//'):
            l.pos += 2
            steps.append(_parse_one_step(l, 'descendant'))
        elif l.peek_str('/'):
            l.pos += 1
            steps.append(_parse_one_step(l, 'child'))
        else:
            break
    return steps


def _parse_one_step(l: _Lexer, axis: str) -> _CXStep:
    l.skip_ws()
    if l.eat_char('*'):
        name = ''
    else:
        name = l.read_ident()
        if not name:
            raise ValueError(f"CXPath parse error: expected element name at pos {l.pos}  expr: {l.src}")
    preds = []
    while True:
        l.skip_ws()
        if l.peek_str('['):
            preds.append(_parse_pred_bracket(l))
        else:
            break
    return _CXStep(axis=axis, name=name, preds=preds)


def _parse_pred_bracket(l: _Lexer) -> Any:
    if not l.eat_char('['):
        raise ValueError(f"CXPath parse error: expected [ at pos {l.pos}  expr: {l.src}")
    l.skip_ws()
    pred = _parse_pred_expr(l)
    l.skip_ws()
    if not l.eat_char(']'):
        raise ValueError(f"CXPath parse error: expected ] at pos {l.pos}  expr: {l.src}")
    return pred


def _parse_pred_expr(l: _Lexer) -> Any:
    left = _parse_pred_term(l)
    l.skip_ws()
    saved = l.pos
    word = l.read_ident()
    if word == 'or':
        l.skip_ws()
        right = _parse_pred_term(l)
        return CXPredBoolOr(left=left, right=right)
    l.pos = saved
    return left


def _parse_pred_term(l: _Lexer) -> Any:
    left = _parse_pred_factor(l)
    l.skip_ws()
    saved = l.pos
    word = l.read_ident()
    if word == 'and':
        l.skip_ws()
        right = _parse_pred_factor(l)
        return CXPredBoolAnd(left=left, right=right)
    l.pos = saved
    return left


def _parse_pred_factor(l: _Lexer) -> Any:
    l.skip_ws()
    # not(...)
    if l.peek_str('not(') or l.peek_str('not ('):
        l.read_ident()  # consume 'not'
        l.skip_ws()
        if not l.eat_char('('):
            raise ValueError(f"CXPath parse error: expected ( after not  expr: {l.src}")
        l.skip_ws()
        inner = _parse_pred_expr(l)
        l.skip_ws()
        if not l.eat_char(')'):
            raise ValueError(f"CXPath parse error: expected ) after not(...)  expr: {l.src}")
        return CXPredNot(inner=inner)
    # contains(@attr, val)
    if l.peek_str('contains('):
        l.read_ident()  # consume 'contains'
        l.skip_ws()
        if not l.eat_char('('):
            raise ValueError(f"CXPath parse error: expected ( after contains  expr: {l.src}")
        l.skip_ws()
        if not l.eat_char('@'):
            raise ValueError(f"CXPath parse error: expected @attr in contains()  expr: {l.src}")
        attr = l.read_ident()
        l.skip_ws()
        if not l.eat_char(','):
            raise ValueError(f"CXPath parse error: expected , in contains()  expr: {l.src}")
        l.skip_ws()
        val = _parse_scalar_str(l)
        l.skip_ws()
        if not l.eat_char(')'):
            raise ValueError(f"CXPath parse error: expected ) after contains(...)  expr: {l.src}")
        return CXPredFuncContains(attr=attr, val=val)
    # starts-with(@attr, val)
    if l.peek_str('starts-with('):
        while l.pos < len(l.src) and l.src[l.pos] != '(':
            l.pos += 1
        if not l.eat_char('('):
            raise ValueError(f"CXPath parse error: expected ( after starts-with  expr: {l.src}")
        l.skip_ws()
        if not l.eat_char('@'):
            raise ValueError(f"CXPath parse error: expected @attr in starts-with()  expr: {l.src}")
        attr = l.read_ident()
        l.skip_ws()
        if not l.eat_char(','):
            raise ValueError(f"CXPath parse error: expected , in starts-with()  expr: {l.src}")
        l.skip_ws()
        val = _parse_scalar_str(l)
        l.skip_ws()
        if not l.eat_char(')'):
            raise ValueError(f"CXPath parse error: expected ) after starts-with(...)  expr: {l.src}")
        return CXPredFuncStartsWith(attr=attr, val=val)
    # last()
    if l.peek_str('last()'):
        l.pos += 6
        return CXPredPosition(is_last=True)
    # (grouped expr)
    if l.peek_str('('):
        l.eat_char('(')
        l.skip_ws()
        inner = _parse_pred_expr(l)
        l.skip_ws()
        if not l.eat_char(')'):
            raise ValueError(f"CXPath parse error: expected ) at pos {l.pos}  expr: {l.src}")
        return inner
    # @attr comparison or existence
    if l.pos < len(l.src) and l.src[l.pos] == '@':
        l.eat_char('@')
        attr = l.read_ident()
        l.skip_ws()
        op = _parse_op(l)
        if not op:
            return CXPredAttrExists(attr=attr)
        l.skip_ws()
        val = _parse_scalar_val(l)
        return CXPredAttrCmp(attr=attr, op=op, val=val)
    # integer position predicate
    if l.pos < len(l.src) and l.src[l.pos].isdigit():
        start = l.pos
        while l.pos < len(l.src) and l.src[l.pos].isdigit():
            l.pos += 1
        return CXPredPosition(pos=int(l.src[start:l.pos]))
    # bare name → child existence
    name = l.read_ident()
    if name:
        return CXPredChildExists(name=name)
    raise ValueError(f"CXPath parse error: unexpected character at pos {l.pos}  expr: {l.src}")


def _parse_op(l: _Lexer) -> str:
    for op in ('!=', '>=', '<=', '=', '>', '<'):
        if l.eat_str(op):
            return op
    return ''


def _autotype_value(s: str) -> Any:
    if s == 'true':  return True
    if s == 'false': return False
    if s == 'null':  return None
    try:    return int(s)
    except ValueError: pass
    try:    return float(s)
    except ValueError: pass
    return s


def _parse_scalar_val(l: _Lexer) -> Any:
    if l.peek_str("'"):
        return l.read_quoted()
    s = l.read_ident()
    if not s:
        raise ValueError(f"CXPath parse error: expected value at pos {l.pos}  expr: {l.src}")
    return _autotype_value(s)


def _parse_scalar_str(l: _Lexer) -> str:
    if l.peek_str("'"):
        return l.read_quoted()
    return l.read_ident()


# ── Evaluator ─────────────────────────────────────────────────────────────────

def collect_step(ctx, expr: _CXPathExpr, step_idx: int, result: list) -> None:
    """Dispatch from context element into its children for the given step."""
    from .ast import Element
    if step_idx >= len(expr.steps):
        return
    step = expr.steps[step_idx]
    if step.axis == 'child':
        candidates = [i for i in ctx.items
                      if isinstance(i, Element) and (step.name == '' or i.name == step.name)]
        for i, child in enumerate(candidates):
            if _preds_match(child, step.preds, candidates, i):
                if step_idx == len(expr.steps) - 1:
                    result.append(child)
                else:
                    collect_step(child, expr, step_idx + 1, result)
    else:
        _collect_descendants(ctx, expr, step_idx, result)


def _collect_descendants(ctx, expr: _CXPathExpr, step_idx: int, result: list) -> None:
    """Descendant axis: match at every depth with proper sibling context for position preds."""
    from .ast import Element
    step = expr.steps[step_idx]
    is_last = step_idx == len(expr.steps) - 1
    candidates = [i for i in ctx.items
                  if isinstance(i, Element) and (step.name == '' or i.name == step.name)]
    for i, child in enumerate(candidates):
        if _preds_match(child, step.preds, candidates, i):
            if is_last:
                result.append(child)
            else:
                collect_step(child, expr, step_idx + 1, result)
        # Always recurse deeper (even after a match) for descendant axis
        _collect_descendants(child, expr, step_idx, result)
    # Also descend into non-matching children for named steps (not needed for wildcard)
    if step.name:
        for child in ctx.items:
            if isinstance(child, Element) and child.name != step.name:
                _collect_descendants(child, expr, step_idx, result)


# ── Predicate evaluators ──────────────────────────────────────────────────────

def _preds_match(el, preds: list, siblings: list, idx: int) -> bool:
    return all(_pred_eval(el, p, siblings, idx) for p in preds)


def _pred_eval(el, pred: Any, siblings: list, idx: int) -> bool:
    if isinstance(pred, CXPredAttrExists):
        return el.attr(pred.attr) is not None
    if isinstance(pred, CXPredAttrCmp):
        v = el.attr(pred.attr)
        if v is None:
            return False
        return _compare(v, pred.op, pred.val)
    if isinstance(pred, CXPredChildExists):
        return el.get(pred.name) is not None
    if isinstance(pred, CXPredNot):
        return not _pred_eval(el, pred.inner, siblings, idx)
    if isinstance(pred, CXPredBoolAnd):
        return (_pred_eval(el, pred.left, siblings, idx)
                and _pred_eval(el, pred.right, siblings, idx))
    if isinstance(pred, CXPredBoolOr):
        return (_pred_eval(el, pred.left, siblings, idx)
                or _pred_eval(el, pred.right, siblings, idx))
    if isinstance(pred, CXPredPosition):
        if pred.is_last:
            return idx == len(siblings) - 1
        return idx == pred.pos - 1
    if isinstance(pred, CXPredFuncContains):
        v = el.attr(pred.attr)
        return v is not None and pred.val in _val_to_str(v)
    if isinstance(pred, CXPredFuncStartsWith):
        v = el.attr(pred.attr)
        return v is not None and _val_to_str(v).startswith(pred.val)
    return False


def _val_to_str(v: Any) -> str:
    if v is None:           return 'null'
    if isinstance(v, bool): return 'true' if v else 'false'
    return str(v)


def _scalar_eq(a: Any, b: Any) -> bool:
    # bool is a subclass of int in Python — guard against cross-type matches
    a_bool = isinstance(a, bool)
    b_bool = isinstance(b, bool)
    if a_bool != b_bool:
        return False
    if isinstance(a, (int, float)) and isinstance(b, (int, float)):
        return float(a) == float(b)
    return a == b


def _compare(actual: Any, op: str, expected: Any) -> bool:
    if op == '=':  return _scalar_eq(actual, expected)
    if op == '!=': return not _scalar_eq(actual, expected)
    a = _to_f64(actual)
    b = _to_f64(expected)
    if op == '>':  return a > b
    if op == '<':  return a < b
    if op == '>=': return a >= b
    if op == '<=': return a <= b
    return False


def _to_f64(v: Any) -> float:
    if isinstance(v, bool):
        raise ValueError(f"CXPath: numeric comparison requires numeric value, got bool: {v}")
    if isinstance(v, (int, float)):
        return float(v)
    raise ValueError(f"CXPath: numeric comparison requires numeric attribute value, got: {v!r}")


# ── cxpath_elem_matches (for transform_all) ───────────────────────────────────

def cxpath_elem_matches(el, expr: _CXPathExpr) -> bool:
    if not expr.steps:
        return False
    last = expr.steps[-1]
    if last.name and last.name != el.name:
        return False
    non_pos = [p for p in last.preds if not isinstance(p, CXPredPosition)]
    return _preds_match(el, non_pos, [], 0)


# ── Transform helpers ─────────────────────────────────────────────────────────

def elem_detached(e) -> Any:
    """Return a copy of e with independent attrs/items lists so f cannot mutate the source."""
    from .ast import Element, Attr
    return Element(
        name=e.name,
        anchor=e.anchor,
        merge=e.merge,
        data_type=e.data_type,
        attrs=[Attr(a.name, a.value, a.data_type) for a in e.attrs],
        items=list(e.items),
    )


def doc_replace_at(d, idx: int, el) -> Any:
    from .ast import Document
    return Document(
        elements=[el if i == idx else n for i, n in enumerate(d.elements)],
        prolog=d.prolog,
        doctype=d.doctype,
    )


def elem_replace_item_at(e, idx: int, child) -> Any:
    from .ast import Element
    return Element(
        name=e.name,
        anchor=e.anchor,
        merge=e.merge,
        data_type=e.data_type,
        attrs=e.attrs,
        items=[child if i == idx else n for i, n in enumerate(e.items)],
    )


def path_copy_element(e, parts: list, f: Callable) -> Optional[Any]:
    """Returns a new Element with f applied at parts[...], or None if path not found."""
    from .ast import Element
    for i, item in enumerate(e.items):
        if isinstance(item, Element) and item.name == parts[0]:
            if len(parts) == 1:
                return elem_replace_item_at(e, i, f(elem_detached(item)))
            updated = path_copy_element(item, parts[1:], f)
            if updated is not None:
                return elem_replace_item_at(e, i, updated)
            return None
    return None


def rebuild_node(node, expr: _CXPathExpr, f: Callable) -> Any:
    """Recursively rebuild node tree, applying f to every element matching expr."""
    from .ast import Element
    if not isinstance(node, Element):
        return node
    new_items = [rebuild_node(item, expr, f) for item in node.items]
    new_el = Element(
        name=node.name,
        anchor=node.anchor,
        merge=node.merge,
        data_type=node.data_type,
        attrs=node.attrs,
        items=new_items,
    )
    if cxpath_elem_matches(new_el, expr):
        return f(elem_detached(new_el))
    return new_el
